[ES|QL] Rewrite In with EvaluatorMapper

fang-xing-esql opened the pull request:
This is a subtask of #106679 to replace EvalMapper with EvaluatorMapper. There are multiple functions/predicates that use EvalMapper, this PR focus on the IN predicate.

The result of In predicate can be different whether the inlist has nulls in it or not. Multi-valued fields/expressions are not support by In predicate.

```
The results below are collected from postgres, these are used as the expected results.
test=# select * from test;
 c1 | c2
----+----
  1 | a
  2 | b
  3 | c
  4 | d
  5 |
(5 rows)

When there is null in the inlist, the return value of in can be either true or null, depending whether there is a matching between LHS and RHS. If LHS is null, in returns null no matter the inlist has null or not.
test=# select c1, c2, c2 in ('e', 'f') as c3 from test;
 c1 | c2 | c3
----+----+----
  1 | a  | f
  2 | b  | f
  3 | c  | f
  4 | d  | f
  5 |    |
(5 rows)

test=# select c1, c2, c2 in ('a', 'f', null) as c3 from test;
 c1 | c2 | c3
----+----+----
  1 | a  | t
  2 | b  |
  3 | c  |
  4 | d  |
  5 |    |
(5 rows)
```

fang-xing-esql committed:
rewrite in with EvaluatorMapper

fang-xing-esql committed:
rewrite inlist

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
rewrite inlist

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
evaluator

fang-xing-esql committed:
EsqlCapabilities

fang-xing-esql commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:
```diff
@@ -68,12 +70,12 @@ public EvaluatorImplementer(
             elements.getPackageOf(declarationType).toString(),
             declarationType.getSimpleName() + extraName + "Evaluator"
         );
+        this.isIn = declarationType.getSimpleName().toString().equalsIgnoreCase("In");
```
There might be a better way to detect whether this is an In predicate?

nik9000 replied:
Could the `blockStyle` invocation work? That's less generated code, but might be ok?

fang-xing-esql replied:
> Could the `blockStyle` invocation work? That's less generated code, but might be ok?

It looks like ArrayProcessFunctionArg does quite work with block type inputs, I'll see if I can make it work and if it is more or less complicated than the current approach.

nik9000 replied:
What about `BlockProcessFunctionArg`? That's kind of "I'll do it manually" for process.

fang-xing-esql replied:
> What about `BlockProcessFunctionArg`? That's kind of "I'll do it manually" for process.

BlockProcessFunctionArg only works block type, it doesn't support block type array... the inlist is an array.

fang-xing-esql committed:
Merge branch 'main' into inlist

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/ip.csv-spec:
```diff
@@ -163,12 +163,11 @@ eth1           |beta           |127.0.0.1
 eth1           |beta           |127.0.0.1                                                                        |128.0.0.1                |128.0.0.1
 lo0            |gamma          |fe80::cae2:65ff:fece:feb9                                                        |fe81::cae2:65ff:fece:feb9|fe81::cae2:65ff:fece:feb9
 eth0           |gamma          |fe80::cae2:65ff:fece:feb9                                                        |127.0.0.3                |127.0.0.3
-eth0           |epsilon        |[fe80::cae2:65ff:fece:feb9, fe80::cae2:65ff:fece:fec0, fe80::cae2:65ff:fece:fec1]|fe80::cae2:65ff:fece:fec1|fe80::cae2:65ff:fece:fec1
```
What's the behavior change here? Do we want that?

fang-xing-esql replied:
If there are multi-valued expressions in the inlist, the test in string.csv-spec returns null, and have a warning says "single-value function encountered multi-value". This test case also has this warning but it doesn't return a null on main, perhaps the old/current implementation handles them(ip and string) differently. I'm trying to make them behave the same.

nik9000 replied:
Oh, interesting! I think @bpintea wrote IN the first time around. @bpintea do you know which behavior we want in all cases?

bpintea replied:
IMO this line should stay: `IN` is syntactic sugar for `==` _and_ `OR`. So the presence of a single MV element in IN's list should not nullify the result: we optimise statements like `a == b OR a == c` as `a IN (b, c)` and the non-IN expression can return true if one equality holds, even when the other is `null` due to comparison against a MV. (Genearlly, this is the case with other functions too, such as `CASE(true, 1, [1, 2])` returning `1`, one MV argument in its argument list doesn't automatically mean a `null`).
Besides that, we need to keep the 3VL: (true, null) -> true, (false, null) -> null.

fang-xing-esql replied:
> IMO this line should stay: `IN` is syntactic sugar for `==` _and_ `OR`. So the presence of a single MV element in IN's list should not nullify the result: we optimise statements like `a == b OR a == c` as `a IN (b, c)` and the non-IN expression can return true if one equality holds, even when the other is `null` due to comparison against a MV. (Genearlly, this is the case with other functions too, such as `CASE(true, 1, [1, 2])` returning `1`, one MV argument in its argument list doesn't automatically mean a `null`). Besides that, we need to keep the 3VL: (true, null) -> true, (false, null) -> null.

Thanks for the explanation! I'll allow MV in the inlist, and also keep the 3VL rule.

not-napoleon commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/ip.csv-spec:
```diff
@@ -146,7 +146,7 @@ fe80::cae2:65ff:fece:fec1                             |[fe80::cae2:65ff:fece:feb
 
 in
-required_capability: mv_warn
+required_capability: in_evaluatormapper
```
My understanding is that, unlike features and transport version, having a capability does not imply having all older capabilities.  So, maybe we want this to require both `mv_warn` and `in_evaluatormapper`?  Although I am just taking it on faith that `required_capability` can accept more than one requirement.

nik9000 replied:
You write `required_capability` twice if you want two.

fang-xing-esql replied:
If the test returns different results between the two `required_capability`, what should I here?

nik9000 replied:
> If the test returns different results between the two required_capability, what should I here?

We're not removing the old capability so you could use both. Or just the one added after. Either would work and I'm not sure I have an opinion yet on which of those two choices is better.

fang-xing-esql replied:
> > If the test returns different results between the two required_capability, what should I here?
> 
> We're not removing the old capability so you could use both. Or just the one added after. Either would work and I'm not sure I have an opinion yet on which of those two choices is better.

I'll replace the old one with the new one for now, as the results changed...

not-napoleon commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java:
```diff
@@ -53,22 +77,84 @@ public boolean foldable() {
+            || commonType == TEXT
+            || commonType == IP
+            || commonType == VERSION
+            || commonType == UNSUPPORTED
```
I'm a little confused as to what would cause `commonType` to be `UNSUPPORTED` (honestly, I'm a little confused as to what the `UNSUPPORTED` type is meant to do in general).  Could you add a comment here to clarify, please?

fang-xing-esql replied:
> I'm a little confused as to what would cause `commonType` to be `UNSUPPORTED` (honestly, I'm a little confused as to what the `UNSUPPORTED` type is meant to do in general). Could you add a comment here to clarify, please?

That's a good question! I think I copied it from Greatest :-), I need to test more to understand the difference between the commonType UNSUPPERED and NULL, perhaps we should return a CONSTANT_NULL_FACTORY for both of them, I'll test it more.

fang-xing-esql committed:
ExpressionTranslator for esql in

fang-xing-esql committed:
update evaluator

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
add BlockArrayProcessFunctionArg to EvaluatorImplementer

fang-xing-esql committed:
Merge branch 'main' into inlist

not-napoleon commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:
```diff
@@ -974,12 +977,114 @@ public String closeInvocation() {
         }
     }
 
+    private record BlockArrayProcessFunctionArg(TypeName componentType, String name) implements ProcessFunctionArg {
```
I know this is private, but I think a line or two of Javadoc about what it's for would still be helpful.

fang-xing-esql replied:
I'll add javadoc.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1814,6 +1815,23 @@ private static LogicalPlan normalize(Aggregate aggregate, AttributeMap<Expressio
 
     public static class FoldNull extends OptimizerRules.FoldNull {
+        @Override
+        public Expression rule(Expression e) {
```
> Is this just copying over the behavior from core? or is this new behavior?

This is just copying over from the core, except that the In is more core's In any more :).

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
added some comments for ProcessFunctionArg

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java:
```diff
@@ -96,7 +201,142 @@ protected TypeResolution resolveType() { // TODO: move the foldability check fro
+        checkMV(p, lhs);
+        Long l = lhs.getLong(lhs.getFirstValueIndex(p));
+        Long[] r = new Long[rhs.length];
+        BitSet nulls = new BitSet(rhs.length);
```
I'd be worried about all of these allocations in `process`. We really want to keep these to a minimum and all of the `Long`, `Long[]` and `BitSet` allocations are quite a lot.

I wonder if we should split this into two tasks - moving to the factory which we can do by writing the factory by hand, and, and reworking these `process` functions which feels a little less important, and harder!

fang-xing-esql replied:
Discussed with Nik, the way that In handles null values and multi-value are quite unique, generate the evaluators through template instead of through annotations.

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
3vl

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
in template

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
generate In evaluators with templates

fang-xing-esql committed:
spotless

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
pick up changes in main

fang-xing-esql committed:
clean up

fang-xing-esql committed:
Merge branch 'main' into inlist

nik9000 commented on x-pack/plugin/esql/src/main/generated-src/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/InBooleanEvaluator.java:
```diff
@@ -0,0 +1,173 @@
+import java.util.BitSet;
+
+/**
+ * This class is generated. Edit {@code X-In.java.st} instead.
```
It'd be nice to have a little javadoc explaining what it does.

nik9000 replied:
Should the generating file be called `InEvaluator.java.st` maybe?

nik9000 commented on x-pack/plugin/esql/src/main/generated-src/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/InBooleanEvaluator.java:
```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    public BooleanBlock eval(int positionCount, BooleanBlock lhsBlock, BooleanBlock[] rhsBlocks) {
```
`private`?

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java:
```diff
@@ -139,4 +168,184 @@ protected TypeResolution resolveType() { // TODO: move the foldability check fro
+        } else {
+            builder.appendBoolean(false);
+        }
+    }
```
Are these worth moving to the generated code?

fang-xing-esql replied:
> Are these worth moving to the generated code?

That's a good idea! I'll move the builder append call into the generated code.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlExpressionTranslators.java:
```diff
@@ -401,4 +404,58 @@ static Query translate(
         }
     }
+
+    public static class InComparisons extends ExpressionTranslator<In> {
```
I think we're moving these into their own top level file more often these days.

fang-xing-esql replied:
> I think we're moving these into their own top level file more often these days.

I just double checked, a lot of `LogicalPlanOptimizer` rules are in their own top level files, under optimizer/rules, the ExpressionTranslators are not in their own files yet. They are in either `EsqlExpressionTranslators` or `ExpressionTranslators` in esql-core. Perhaps keep it here for now?

nik9000 replied:
Either is fine by me.

nik9000 commented on x-pack/plugin/esql/src/main/generated-src/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/InBooleanEvaluator.java:
```diff
@@ -0,0 +1,173 @@
+                    continue;
+                }
+                try {
+                    In.process(result, nulls, mvs, lhsBlock.getBoolean(lhsBlock.getFirstValueIndex(p)), rhsValues);
```
I wonder - should `boolean`s be special and not use the array method? Ultimately `IN (true)` and `IN (false)` and `IN ()` and `IN (false, true)` are the only possible layouts. Right? `IN (false, false) has the same meaning as `IN (false)`. I think.

If that's true, then this is useful: I've found sometimes I don't use the string template stuff for `boolean` because it's so different. In this case maybe I'd slurp more of this method into `In.process`'s boolean flavor. But whatever make the hacking easy I'm for. I don't have a strong opinion about the right way to lay out these templates.

fang-xing-esql replied:
Indeed, I'll simplify boolean.

nik9000 commented on x-pack/plugin/esql/src/main/generated-src/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/InBytesRefEvaluator.java:
```diff
@@ -0,0 +1,185 @@
+                }
+                try {
+                    In.process(result, nulls, mvs, lhsBlock.getBytesRef(lhsBlock.getFirstValueIndex(p), lhsScratch), rhsValues);
+                } catch (IllegalArgumentException e) {
```
Are we throwing this one intentionally? I think maybe we can just not do the try/catch here because we really shouldn't be throwing this. I think.

fang-xing-esql replied:
> Are we throwing this one intentionally? I think maybe we can just not do the try/catch here because we really shouldn't be throwing this. I think.

I'll remove the two try/catch, seems like not needed and I don't recall why they are there in the first place, perhaps the `process` method changed.

nik9000 reviewed (commented):
I like this much more! I left a few questions, but I think this seems like the right thing to do for `IN`.

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
update according to review comments

fang-xing-esql committed:
cleanup

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
cleanup

fang-xing-esql commented:
> I like this much more! I left a few questions, but I think this seems like the right thing to do for `IN`.

Thanks for reviewing it again @nik9000 ! Most of the comments are addressed, and I cleaned it up according to the suggestions, simplified boolean a bit, left `InComparisons` in `EsqlExpressionTranslators` for now as the other `ExpressionTranslator`s.

fang-xing-esql committed:
Merge branch 'main' into inlist

fang-xing-esql committed:
Merge branch 'main' into inlist

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -929,15 +929,15 @@ public LogicalPlan apply(LogicalPlan plan, AnalyzerContext context) {
             }
-            if (f instanceof In in) {
-                return processIn(in);
-            }
```
Can we use `processScalarFunction` for this one or is it incompatible somehow?

fang-xing-esql replied:
> Can we use `processScalarFunction` for this one or is it incompatible somehow?

This is a good point! It is not quite compatible for now, one reason is that In is not a registered function yet, it is under operator/comparison, there will be a little more work to make it a registered function. I'll look into it, and if possible, create a follow up PR for it.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/InEvaluator.java.st:
```diff
@@ -0,0 +1,273 @@
+import org.elasticsearch.compute.data.BytesRefVector;
+$elseif(boolean)$
+import org.elasticsearch.compute.data.BooleanVector;
+$endif$
```
I've tried to remove the import order rules from this generated code but it's a huge pain. I'm sorry about this mess.

nik9000 reviewed (approved):
Looks good to me. I left a question about the Analyzer but that's pretty far outside my area of expertise so I'm happy to trust about moving that invocation rather than merging.
